src/join_reference.rs

```diff
@@ -0,0 +1,89 @@
+use std::fmt::{Debug, Display, Formatter};
+
+use flexstr::SharedStr;
+use serde::{Deserialize, Serialize};
+
+use crate::reference::Reference;
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[repr(transparent)]
+pub struct JoinReference(Reference);
+impl JoinReference {
+    pub fn new() -> Self {
+        Self(Reference::new())
+    }
+}
+impl Debug for JoinReference {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        if f.alternate() {
+            write!(f, "{:#?}", self.0)
+        } else {
+            write!(f, "JoinReference({:?})", self.0)
+        }
+    }
+}
+impl Display for JoinReference {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+impl From<&str> for JoinReference {
```

Comment:
From<&str> and From<String> can be consolidated with:

impl T From<T> for JoinReference where T: Into<Reference>